File: CourseProject/ILS_WPF/Models/Core/WarehouseEntry.cs
using ILS_WPF.Models.Core.Enums;
using ILS_WPF.Models.Core.Requests;
using ILS_WPF.Models.Database;

namespace ILS_WPF.Models.Core
{
    public abstract class WarehouseEntry<T> : IDbEntry
    {
        public int Id { get; set; }
        public string? Name { get; set; }
        public MaterialType MaterialType { get; set; }
        public int? WarehouseId { get; set; }
        public virtual Warehouse? Warehouse { get; set; }

        public abstract bool IsEmpty();
        public abstract void Increase(T amount);
        public abstract T Decrease(T amount);
        public abstract bool IsMatches(SupplyRequestDetail detail);
    }
}

File: lab8/Java/src/warehouse/WarehouseEntry.java
package warehouse;

import enums.MaterialType;
import request.SupplyRequestDetail;

public abstract class WarehouseEntry<T extends Number> {

    String name;
    MaterialType materialType;
    protected int id;
    protected int assignedWarehouseId;
    protected Warehouse assignedWarehouse;

    public WarehouseEntry(int id, String name, MaterialType materialType){
        this.id = id;
        this.name = name;
        this.materialType = materialType;
    }

    public int getId(){
        return id;
    }
    public String getName(){
        return name;
    }
    public MaterialType getMaterialType(){
        return materialType;
    }
    public void assignWarehouse(Warehouse warehouse){
        if(warehouse!=null){
            this.assignedWarehouseId = warehouse.getId();
            this.assignedWarehouse = warehouse;
        }
    }
    public abstract Boolean isMatches(SupplyRequestDetail detail);
    public abstract Boolean isEmpty();
    public abstract void increase(Number v);
    public abstract T decrease(Number v);
}
